Don't auto add packages if they cause conflicts; Prompt for conflicts BEFORE adding a package

guillaumejenkins opened the pull request:
Fixes #441
Fixes #905
Fixes #1119
Fixes #1159
Fixes #1232
Fixes #1236

guillaumejenkins commented:
New experience:
- Missing packages are no longer added automatically if they cause a conflict with existing packages
- When adding a package, we detect any conflicts with existing packages, and warn *before* installing it
    - If no conflicts are detected, the new package is added as usual
    - If a conflict is detected, we try to detect if the conflicting packages are in use
        - If the conflicting packages are *not* used anywhere, then we warn the user that they will be removed (with the ability to cancel).
[Screenshot coming]
        - If the conflicting packages appear to be in use in the user's blocks, then we let them know that they need to remove all these blocks before they can add the new package
[Screenshot coming]

pelikhan commented:
Could we try to get some test cases for this logic? It's getting quite intricate.

pelikhan commented on webapp/src/app.tsx:
```diff
@@ -260,14 +258,71 @@ class ScriptSearch extends data.Component<ISettingsProps, ScriptSearchState> {
+                                header: lf("Packages will be removed"),
+                                agreeLbl: lf("Remove packages and add {0}", config.name),
+                                agreeClass: "pink",
+                                htmlBody: `<p>${lf("The following package(s) are incompatible with {0} and will be removed. Continue?", config.name, conflictNames)}<br/><br/>${conflictNames}`
```
Use "body" or escape all user provided strings.

pelikhan commented on pxtlib/main.ts:
```diff
@@ -189,15 +189,33 @@ namespace pxt {
+            return Promise.resolve().then(() => {
+                if (ghConfig) {
+                    return pxt.packagesConfigAsync()
+                        .then(config => pxt.github.latestVersionAsync(ghConfig[1], config))
```
not sure why we call lastestVersionAsync since we do not use ``tag`` afterwards

pelikhan commented on pxtlib/main.ts:
```diff
@@ -310,13 +348,84 @@ namespace pxt {
+                    const additionalConflicts: cpp.PkgConflictError[] = [];
+                    conflicts.forEach((c) => {
+                        additionalConflicts.push.apply(additionalConflicts, allAncestors(c.pkg0).map((anc) => {
+                            const confl = new cpp.PkgConflictError("");
```
no details?

pelikhan commented on pxtlib/main.ts:
```diff
@@ -310,13 +348,84 @@ namespace pxt {
+                        return true;
+                    });
+                })
+                .then(() => conflicts);
```
when `.then(() `` here?

pelikhan commented on webapp/src/app.tsx:
```diff
@@ -260,14 +258,71 @@ class ScriptSearch extends data.Component<ISettingsProps, ScriptSearchState> {
+                                header: lf("Cannot add {0} package", config.name),
+                                hideCancel: true,
+                                agreeLbl: lf("Ok"),
+                                htmlBody: lf("Remove all the blocks from the {0} package and try again.", inUse.pkg0.id)
```
Use "body" instead.

pelikhan commented:
![image](https://cloud.githubusercontent.com/assets/4175913/22234100/411cbe2e-e1aa-11e6-8923-648bc764e19b.png)

Make it a single sentence: Packages ... are incompatiable with blah blah blah...

pelikhan commented:
The simulator does not seem to restart after adding a package.

guillaumejenkins commented:
The simulator issue was because I hadn't merged develop in a while; I've done that now and addressed feedback, as well as added a few comments to improve clarity

Next up I'll add some tests, at least for the functions that I've added in pxtlib

pelikhan commented:
We should try to add end-to-end integration. Maybe you can add a CLI command that allows to add a package (eg "pxt install --save") so that we can automate test cases for the logic.
